### PR TITLE
Cudnn v4

### PR DESCRIPTION
**Pull request by kashif**
Added support for cudnn v4.

### CODE REVIEW

**Comment by unnonouno** [comment]
Thank you for a patch. Are these method not supported on cuDNN v2?

**Comment by kashif** [comment]
@unnonouno no the cudnn v2 api is not similar and many of the methods have been deprecated by v3 and v4 of cudnn. I think if possible we should also switch over to v4 of the api. What do you think?

**Comment by unnonouno** [comment]
Thank you for information. I think new APIs are better on v3 and v4 as you wrote. But, we also want to support v2 because it is not old enough. v2 is only one year old.

**Comment by beam2d** [comment]
There are two levels of support for cuDNN v4.

- Support environments with cuDNN v4 installed. In this case, we can use ***_v2 APIs to support v2, v3, and v4 at once.
- Support full capability of installed cuDNN version. In this case, we have to switch ***_v2 APIs and ***_v4 APIs based on the installed cuDNN version.

We will add the former support by another PR. After the PR is merged, would you merge it and add a v2 support to this PR here based on the former support? I recommend to use APIs with suffixes rather than non-suffix version (which is inconsistent between v2/v3 and v4).

**Comment by kashif** [comment]
yes sure @beam2d please let me know when the PR is merged and i will then make the changes. Thank you!

**Comment by okuta** [comment]
I made the PR #838 that support environments with cuDNN v4 installed.

**Comment by kashif** [comment]
thank you @okuta I will now add my changes on top of this...

**Comment by okuta** [comment]
@kashif Do you have time to work on this PR? Since it requires some knowledge to resolve conflicts, we will help you to fix the code.

We would like to merge this pull request in v1.7.
I am sorry to make you hurry, but we will work on sending you PR if no response is returned until the end of this week (2/12).

**Comment by kashif** [comment]
so sorry @okuta I will send you the fixed pull request soon. thanks!

**Comment by okuta** [comment]
OK, We are waiting for the fixed pull request.

**Comment by kashif** [comment]
@okuta ok i have merged my changes into master and next I will need to seperate out the v2/v3 and v4 functions. Should i do it via `#if CUDNN_VERSION` definitions?

**Comment by okuta** [comment]
Yes, Please use `#if CUDNN_VERSION` deginitions.
For your information https://github.com/pfnet/chainer/blob/master/cupy/cuda/cupy_cudnn.h#L11 .

**Comment by kashif** [comment]
@okuta can you kindly review? Thank you!

**Comment by okuta** [comment]
Thank you! I will review this.

File: README.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ Requirements for some features:
   - filelock
   - g++ 4.8.4+
 - cuDNN support
-  - cuDNN v2, v3
+  - cuDNN v3, v4
```
Comment (okuta) [inline README.md]: Please add `v2`.

File: cupy/cuda/cudnn.pxd [hunk 2]
```diff
@@ -9,8 +9,13 @@ cdef extern from *:
     ctypedef int AddMode 'cudnnAddMode_t'
     ctypedef int ConvolutionFwdAlgo 'cudnnConvolutionFwdAlgo_t'
     ctypedef int ConvolutionFwdPreference 'cudnnConvolutionFwdPreference_t'
+    ctypedef int ConvolutionBwdFilterPreference 'cudnnConvolutionBwdFilterPreference_t'
```
Comment (okuta) [inline cupy/cuda/cudnn.pxd]: Please sort in lexical order.

File: cupy/cuda/cupy_cudnn.h [hunk 3]
```diff
@@ -32,8 +32,13 @@ typedef int cudnnActivationMode_t;
 typedef int cudnnAddMode_t;
 typedef int cudnnConvolutionFwdAlgo_t;
 typedef int cudnnConvolutionFwdPreference_t;
+typedef int cudnnConvolutionBwdFilterPreference_t;
```
Comment (okuta) [inline cupy/cuda/cupy_cudnn.h]: Please sort in lexical order.

File: cupy/cuda/cupy_cudnn.h [hunk 4]
```diff
@@ -44,6 +49,9 @@ typedef int ActivationMode;
 typedef int AddMode;
 typedef int ConvolutionFwdAlgo;
 typedef int ConvolutionFwdPreference;
+typedef int ConvolutionBwdFilterAlgo;
```
Comment (okuta) [inline cupy/cuda/cupy_cudnn.h]: Please sort in lexical order.

File: cupy/cudnn.py [hunk 5]
```diff
@@ -47,8 +47,10 @@ def get_data_type(dtype):
         return cudnn.CUDNN_DATA_FLOAT
     elif dtype.type == numpy.float64:
         return cudnn.CUDNN_DATA_DOUBLE
+    elif dtype.type == numpy.float16:
+        return cudnn.CUDNN_DATA_HALF
     else:
-        raise TypeError('Dtype {} is not supported in CuDNN v2'.format(dtype))
+        raise TypeError('Dtype {} is not supported in CuDNN'.format(dtype))
```
Comment (okuta) [inline cupy/cudnn.py]: Also, can you fix typo? `CuDNN` => `cuDNN`

**Comment by okuta** [comment]
Please fix tests to check that calling the function of cuDNN.

https://github.com/pfnet/chainer/blob/859cf4c97c1755f7bdd9ed13065026881e825e68/tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py#L170
https://github.com/pfnet/chainer/blob/859cf4c97c1755f7bdd9ed13065026881e825e68/tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py#L161
https://github.com/pfnet/chainer/blob/23c7b8190f8129dabb0e51f6947b1782cd333dda/tests/chainer_tests/links_tests/connection_tests/test_mlp_convolution_2d.py#L85

**Comment by kashif** [comment]
@okuta can you kindly check?

**Comment by okuta** [comment]
LGTM!

**Comment by okuta** [comment]
Sorry, build of this PR fails with cuDNN v2 and v3.
please add follow stub functions to `if CUDNN_VERSION < 4000` section.
```
cudnnConvolutionBackwardData_v3
cudnnGetConvolutionBackwardDataWorkspaceSize
cudnnGetConvolutionBackwardDataAlgorithm
cudnnConvolutionBackwardFilter_v3
cudnnGetConvolutionBackwardFilterWorkspaceSize
cudnnGetConvolutionBackwardFilterAlgorithm
```

**Comment by okuta** [comment]
Sorry, build of this PR fails with cuDNN v2 and v3.
@kashif please add follow stub functions to `if CUDNN_VERSION < 4000` section.
```
cudnnConvolutionBackwardData_v3
cudnnGetConvolutionBackwardDataWorkspaceSize
cudnnGetConvolutionBackwardDataAlgorithm
cudnnConvolutionBackwardFilter_v3
cudnnGetConvolutionBackwardFilterWorkspaceSize
cudnnGetConvolutionBackwardFilterAlgorithm
```

**Comment by okuta** [comment]
Thnak you!
LGTM!

**Comment by kashif** [comment]
great thank you @okuta :bow:

**Pull request merged**